File: Back-End/Back-End/hospital-management-system-main/src/main/java/com/hospital/hms/service/NotificationService.java
package com.hospital.hms.service;

import com.hospital.hms.Enum.NotificationType;
import com.hospital.hms.dto.NotificationDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface NotificationService {

    /** Create and persist a notification for a specific user */
    void sendNotification(Long recipientId, String title, String message,
                          NotificationType type, String actionUrl);

    /** Get all notifications for a user (paginated) */
    Page<NotificationDTO> getNotifications(Long recipientId, Pageable pageable);

    /** Get all notifications for a user (non-paginated, for full page view) */
    List<NotificationDTO> getAllNotifications(Long recipientId);

    /** Get only unread notifications */
    List<NotificationDTO> getUnreadNotifications(Long recipientId);

    /** Count unread notifications */
    long countUnread(Long recipientId);

    /** Mark a single notification as read */
    NotificationDTO markAsRead(Long notificationId);

    /** Mark all notifications as read for a user */
    int markAllAsRead(Long recipientId);

    /** Delete a notification */
    void deleteNotification(Long notificationId);
}

File: Front-End/src/services/notificationService.ts
import api from '@/lib/api';

export interface NotificationDto {
  id: number;
  title: string;
  message: string;
  type: string;
  read: boolean;
  actionUrl?: string;
  createdAt: string;
  readAt?: string;
  recipientId: number;
}

const notificationService = {
  /** Paginated — used by badge dropdown */
  async getAll(userId: number | string, page = 0, size = 20): Promise<{ content: NotificationDto[]; totalElements: number }> {
    const { data } = await api.get(`/api/notifications/user/${userId}`, { params: { page, size } });
    return {
      content: data.content ?? [],
      totalElements: data.page?.totalElements ?? data.totalElements ?? 0,
    };
  },

  /** Non-paginated full list — used by the Notifications page */
  async getAllList(userId: number | string): Promise<NotificationDto[]> {
    const { data } = await api.get<NotificationDto[]>(`/api/notifications/user/${userId}/all`);
    return data;
  },

  async getUnread(userId: number | string): Promise<NotificationDto[]> {
    const { data } = await api.get<NotificationDto[]>(`/api/notifications/user/${userId}/unread`);
    return data;
  },

  async countUnread(userId: number | string): Promise<number> {
    const { data } = await api.get<{ count: number }>(`/api/notifications/user/${userId}/unread/count`);
    return data.count;
  },

  async markAsRead(notificationId: number | string): Promise<NotificationDto> {
    const { data } = await api.put<NotificationDto>(`/api/notifications/${notificationId}/read`);
    return data;
  },

  async markAllAsRead(userId: number | string): Promise<number> {
    const { data } = await api.put<{ updated: number }>(`/api/notifications/user/${userId}/read-all`);
    return data.updated;
  },

  async delete(notificationId: number | string): Promise<void> {
    await api.delete(`/api/notifications/${notificationId}`);
  },
};

export default notificationService;
